### ISSUE TITLE
Query: Can Include be a no-op on non-EF queries

### ISSUE THREAD

**Issue by AlexBilonog** [aspnet/EntityFramework#5735]
Cache pattern fails: (in EF 6 works fine)
```
using (var db = new TheContext())
{
   IQueryable<User> queryable;
   if (false) // if it's a single call
   {
      queryable = db.Set<User>().AsQueryable();
   }
   else // it's a batch mode: it needs to cache users
   {
      queryable = db.Set<User>()
         .Include(r => r.UserDepartments)
         .ThenInclude(r => r.Department)
         .ToList()
         .AsQueryable();
   }

   // ... passing 'queryable' as a parameter to class #2

   // In class #2: there are one or several queries like this:
   var user = queryable
      .Include(r => r.UserDepartments)
      .ThenInclude(r => r.Department)
      .First(r => r.ID == id); // <- exception

   // ... some logic
  }
```
```
An unhandled exception of type 'System.InvalidOperationException' occurred in System.Core.dll

Additional information: There is no method 'Include' on type 'Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions' that matches the specified arguments

Stack trace:
   at System.Linq.EnumerableRewriter.FindMethod(Type type, String name, ReadOnlyCollection`1 args, Type[] typeArgs, BindingFlags flags)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.OldExpressionVisitor.Visit(Expression exp)
   at System.Linq.Expressions.OldExpressionVisitor.VisitExpressionList(ReadOnlyCollection`1 original)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.OldExpressionVisitor.Visit(Expression exp)
   at System.Linq.Expressions.OldExpressionVisitor.VisitExpressionList(ReadOnlyCollection`1 original)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.OldExpressionVisitor.Visit(Expression exp)
   at System.Linq.EnumerableExecutor`1.Execute()
   at System.Linq.EnumerableQuery`1.System.Linq.IQueryProvider.Execute[S](Expression expression)
   at System.Linq.Queryable.First[TSource](IQueryable`1 source, Expression`1 predicate)
   at EFGetStarted.ConsoleApp.Program.Main() in C:\Work\Console4\EFGetStarted.ConsoleApp\Program.cs:line 30
   at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
   at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

EF Core version: 1.0.0-rc3-21338
Operating system: Windows 10
Visual Studio version: VS 2015 last updates

**Comment by ilmax** [comment]
Why you are materializing you query and the using ThenInclude? That don't sound right to me

**Comment by AlexBilonog** [comment]
Firstly I updated my issue.Secondly I will explain more details and why it's needed:

Case 1. Class #2 has to process several users or thousands of users. In the second case it will be more efficient to cache the users (to get the normal performance). As an option I could have implemented separate queries (that work with collections in memory and don't use Include() or ThenInclude()). But then I would duplicate all my queries with many if-else. But I try to avoid additional complexity.

Case 2. I have 2 classes that need to process thousands of users (for example Validator class and Updater class). Their logic uses queries to User table (with approximately the same set of included navigation properties). Instead of caching the users twice I pass to classes single cached in memory collection (with all the included navigation properties). Again I try to avoid additional complexity and don't want to have 2 versions of my queries (1 works with IQueryable, 2 works with IEnumerable)

P.S. A have a project that worked on EF6. I migrated it to EF Core. As temporary workaround I have extension method "IncludeEf6" that fully emulates old one. If it sees that it's not a DB IQueryable, then just returns.

But I would want a 'standard' implementation of this powerful feature in EF Core.
